### DISCUSSION TITLE
Bug: file ignored in flat config is still parsed by the parser

### DISCUSSION THREAD

**Discussion by timmywil** [eslint/eslint#17412] [Q&A] [answered]
### Environment

Node version: v18.16.1
npm version: v9.5.1
Local ESLint version: v8.45.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.19045

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
  files: [
    "test/**"
  ],
  ignores: [
    "test/data/json_obj.js"
  ]
}
```
</details>

```js
{ "data": {"lang": "en", "length": 25} }
```

I'm currently in the process of migrating jQuery's fairly complicated eslint config to use flat config instead. Other than some issues I'm having with the import plugin, this is the only issue left. I want to avoid changing the test or JS file, which means I need eslint not to parse it at all.

### What did you expect to happen?

I don't think the JS should still be parsed. The file can be correctly ignored in the config, but the parser is still parsing it and throwing a syntax error. There are valid use cases for ignoring invalid JS, bypassing the parser altogether, and currently no **flat** configuration that I've tried will fix that. Besides, it's also a performance concern. Seems like a waste of resources to parse a file that isn't going to be linted.

### What actually happened?

```sh
error  Parsing error: Unexpected token :
```

Changing the file to be valid JS does make it clear the file is still ignored by the linter, so `ignores` is set properly, but I don't want it to even try to parse the file. This worked fine before flat config.

### Link to Minimal Reproducible Example

https://github.com/timmywil/eslint-test

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

I was able to work around it for now, by setting the parser to return an empty AST.

```js
{
  files: [ "test/data/json_obj.js" ],
  languageOptions: {
    parser: {
      meta: {
        name: "Ignore Without Parsing"
      },

      // Ignore invalid JS
      parse: function() {
        return {
          type: "Program",
          loc: {},
          comments: [],
          range: [
            0,
            0
          ],
          body: [],
          tokens: []
        };
      }
    }
  }
}
```

**Comment by aladdin-add** [comment]
this is working as expected - as explained in https://github.com/eslint/eslint/issues/17400#issuecomment-1646543502.

**Reply by timmywil** [reply]
Sorry, I'm not seeing how that comment addresses this issue. It corrects the OP on how to ignore a file, but does not explain why the file is still parsed.

**Comment by timmywil** [comment]
Please note this is not an issue about "ignores not working" or a file still being linted unexpectedly. When a file is properly ignored and NOT linted, I'm asking why is the file still parsed by espree (or whatever parser the user is using). The linked comment does not address that. This has been moved to a discussion so you seem confident it's correct. Fair enough, maybe it's intended. But by simply linking to a comment that is unrelated, it seems like you don't understand this issue. Perhaps I threw you off because I originally thought #17400 was related from the title and commented there, but it was not and I have since deleted that comment and made a new issue.

**Reply by aladdin-add** [reply, marked as answer]
Yes, I understand. It's designed to work in this way.

* If it has other keys specified, it means to apply this configuration item to files other than this one.
* it's truly ignored (won't be parsed) if `ignores` in a configuration object that has no other keys specified. You can consider it an exception.

**Comment by timmywil** [comment]
Wait, ok, I could have sworn I tried global `ignores`, but I guess I had something wrong. So, the point is that only the global ignores bypasses the parser. Sorry that took so long for me to click. I still think it's weird `ignores` with `files` still parses the file, but using global `ignores` is at least more elegant than overriding the parser to return an empty AST. I have noticed in researching other issues that the global `ignores` works a bit differently in other ways as well from `ignores` with `files`. Again, I still find the current behavior confusing, but at this point I think I've lost all credibility being so sure of myself and wrong for too long.

**Comment by ivodolenc** [comment]
When you think about it, this current behavior isn't actually a bad thing, there are times when you want to ignore configuration for specific files and when you want to ignore files completely so these are different "ignore modes".

But it's definitely not a good name choice for those options because I imagine more people will complain about it because one option have multiple "ignore modes" so this is confusing.

It might not be a bad idea to separate them by name so you know exactly which option does which job.

For example:

```js
export default [
  {
    files: [...],
    // the behavior of the option remains the same, only the name changes
    ignoreConfig: [...]
  }
]
```

```js
export default [
  {
    files: [...]
  },
  {
    // the behavior of the option remains the same, only the name changes
    ignorePattern: [...] // or ignores
  }
]
```

That way, you know exactly what a certain ignore "mode" does, and it's easier to search the docs if something isn't clear.

What do you think?

**Comment by timmywil** [comment]
I think the ideal API would have `ignores` with `files` simply do the same thing as global `ignores`, but I understand that the cascading behavior of the config might make that difficult. The idea of different names is good if the current behavior is intended for the long run. It would help get across the different ignore modes, as you put it. Otherwise, this will keep coming up in future issues.

**Reply by ivodolenc** [reply]
> Otherwise, this will keep coming up in future issues.

Exactly..
